### PR TITLE
feat(stream): use memcomparable row as Cache key in GenericExtremeState

### PR DESCRIPTION
**Pull request by Gun9niR**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

`OrderedRow` is a mem-consuming struct, as the size of the underlying `OrderedDatum` is that of the longest variant. So we use memcomparable as the key of the cache to reduce memory consumption.

## Checklist

- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

#4884

### CODE REVIEW

**Review by richardchien — APPROVED** [review]
LGTM!

File: src/stream/src/executor/managed_state/aggregation/extreme.rs [hunk 1]
```diff
@@ -154,13 +157,14 @@ impl<S: StateStore> GenericExtremeState<S> {
...
     }
 
-    fn state_row_to_cache_entry(&self, state_row: &Row) -> (OrderedRow, Datum) {
-        let cache_key = OrderedRow::new(
+    fn state_row_to_cache_entry(&self, state_row: &Row) -> StreamExecutorResult<(Vec<u8>, Datum)> {
+        let ordered_row = OrderedRow::new(
             state_row.by_indices(&self.state_table_order_col_indices),
             &self.state_table_order_types,
         );
+        let cache_key = ordered_row.serialize()?;
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/aggregation/extreme.rs:165]: How about directly serializing here?

Comment (Gun9niR) [reply src/stream/src/executor/managed_state/aggregation/extreme.rs:165]: Sure can. I just wanted to keep the code concise.

**Comment by Gun9niR** [comment]
Here's the benchmark result. I've run Nexmark Q5 for 1h30min on a c5.4xlarge machine.

Memory usage: main (above) vs. this PR (below) = 38MB vs. 8MB

main

memcomp

---

To rule out the possibility that the throughput has dropped, I checked the memory usage of HashJoin, and they are basically the same.

(main vs. this PR)

main-join

memcom-join

File: src/stream/src/executor/managed_state/aggregation/extreme.rs [hunk 2]
```diff
@@ -154,13 +157,21 @@ impl<S: StateStore> GenericExtremeState<S> {
...
+            .iter()
+            .zip_eq(self.state_table_order_types.iter())
+            .try_for_each(|(idx, order_type)| {
+                serializer.set_reverse(match order_type {
+                    OrderType::Ascending => false,
+                    OrderType::Descending => true,
+                });
+                serialize_datum_into(&(state_row.0)[*idx], &mut serializer)
+            })?;
+        let cache_key = serializer.into_inner();
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/aggregation/extreme.rs:172]: Can we use `OrderedRowSerializer::serialize_datums` here?

File: src/stream/src/executor/managed_state/aggregation/extreme.rs [hunk 3]
```diff
@@ -154,13 +157,17 @@ impl<S: StateStore> GenericExtremeState<S> {
         }
     }
 
-    fn state_row_to_cache_entry(&self, state_row: &Row) -> (OrderedRow, Datum) {
-        let cache_key = OrderedRow::new(
-            state_row.by_indices(&self.state_table_order_col_indices),
-            &self.state_table_order_types,
+    fn state_row_to_cache_entry(&self, state_row: &Row) -> StreamExecutorResult<(Vec<u8>, Datum)> {
+        let mut cache_key = Vec::new();
+        let serializer = OrderedRowSerializer::new(self.state_table_order_types.clone());
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/aggregation/extreme.rs:162]: That'll be costly as we clone the order types every time this function is called. Looks like the `state_table_order_types` is only used here, I think it's better to simply store an instance of `OrderedRowSerializer` in self.

**Review by BugenZhao — APPROVED** [review]
LGTM

**Pull request merged**